Introduce loader_kwargs [v2]

KennethEnevoldsen opened the pull request:
- removed all partials from loader args and moved them to new arg, loader_kwargs
- explicitly added the loader as sentence_transformers_loader, now None is not impl. - as a consequence I removed the `no_model_implementation_available` implementation.
- fixed similarity_fn_name to Enum where I saw it
- fixed typing issue where I saw it
- removed case where we fetch the distance function for sentence transformers (as this can be more than the 3 we implement and thus lead to a unintended error)

fixes  #1779

Also:
- Should remove the bedrock version of cohere models @Samoed (I am unsure why we have them implemented)


<!-- If you are submitting a dataset or a model for the model registry please use the corresponding checklists below otherwise feel free to remove them. -->

<!-- add additional description, question etc. related to the new dataset -->


### Code Quality
<!-- Please do not delete this -->
- [x] **Code Formatted**: Format the code using `make lint` to maintain consistent style.

### Documentation
<!-- Please do not delete this -->
- [x] **Updated Documentation**: Add or update documentation to reflect the changes introduced in this PR.

### Testing
<!-- Please do not delete this -->
- [ ] **New Tests Added**: Write tests to cover new functionality. Validate with `make test-with-coverage`.
- [x] **Tests Passed**: Run tests locally using `make test` or `make test-with-coverage` to ensure no existing functionality is broken.

Samoed commented on mteb/model_meta.py:
```diff
@@ -128,10 +121,10 @@ class ModelMeta(BaseModel):
-        sentence_transformers uses Literal['cosine', 'dot', 'euclidean', 'manhattan'] for similarity_fn_name
-        pylate uses Literal['MaxSim'] for similarity_fn_name
+        sentence_transformers uses Literal['cosine', 'dot', 'euclidean', 'manhattan'] for
+        pylate uses Literal['MaxSim'] for
```
```suggestion
        pylate uses Literal['MaxSim']
```

Samoed commented on mteb/model_meta.py:
```diff
@@ -128,10 +121,10 @@ class ModelMeta(BaseModel):
     @field_validator("similarity_fn_name", mode="before")
     @classmethod
-    def validate_similarity_fn_name(cls, value):
+    def validate_(cls, value):
```
Why function name was changed?

Samoed reviewed (approved):
Awesome changes!

KennethEnevoldsen commented on mteb/models/nomic_models_vision.py:
```diff
@@ -161,11 +160,7 @@ def get_fused_embeddings(
 nomic_embed_vision_v1_5 = ModelMeta(
-    loader=partial(
-        NomicVisionModelWrapper,
-        vision_model_name="nomic-ai/nomic-embed-vision-v1.5",
```
fixed

KennethEnevoldsen commented:
Fixed repllama, comments, and tests. Does the model_loading script test all models?

Samoed commented on mteb/models/cohere_models.py:
```diff
@@ -217,9 +216,8 @@ def encode(
 cohere_mult_3 = ModelMeta(
-    loader=partial(  # type: ignore
-        CohereTextEmbeddingModel,
-        model_name="embed-multilingual-v3.0",
```
Cohere models needs name without `Cohere/Cohere-` to work correctly. Maybe add to `__init__` `model_name..lstrip("Cohere/Cohere-")`?

KennethEnevoldsen replied:
Ahh good catch - fixed

Samoed commented on mteb/models/openai_models.py:
```diff
@@ -122,9 +121,8 @@ def _to_numpy(self, embedding_response) -> np.ndarray:
     languages=None,  # supported languages not specified
-    loader=partial(  # type: ignore
-        OpenAIWrapper,
-        model_name="text-embedding-3-small",
```
Similar issue with openai

KennethEnevoldsen commented:
@Samoed discovered that the `MockTorchbf16Encoder` test doesn't pass (which was hidden due to it being wrongly set as a sentencetransformer and thus wrapped). We do indeed not support bf16, but I have changed the test to fp16 (without sentence transformers). I have also added a test that ensures that invalid formats from the sentence transformers are converted to a valid format.
